### PR TITLE
GH-101357: Suppress `OSError` from `pathlib.Path.exists()` and `is_*()`

### PR DESCRIPTION
**Pull request by barneygale**
Suppress all `OSError` exceptions from `pathlib.Path.exists()` and `is_*()` rather than a selection of more common errors as we do presently. Also adjust the implementations to call `os.path.exists()` etc, which are much faster on Windows thanks to GH-101196.

* Issue: gh-101357

----
📚 Documentation preview 📚: https://cpython-previews--118243.org.readthedocs.build/

### CODE REVIEW

**Comment by nineteendo** [comment]
What do you intend to do with #117858? This pull request includes all its changes (except for the news entry).

**Comment by zooba** [comment]
Is it worth pointing out that this brings them into line with the `os.path` equivalents? And maybe suggesting the right API (I assume `stat`/`lstat`) to get more detailed error.

There might also be value in some kind of conceptual comment that these functions mean what they say if they return `True`, but could mean multiple things if they return `False`, and hence if you're using them you should be testing for `True`?

**Comment by barneygale** [comment]
Thanks for the feedback!

> There might also be value in some kind of conceptual comment that these functions mean what they say if they return `True`, but could mean multiple things if they return `False`, and hence if you're using them you should be testing for `True`?

It seems quite redundant to me, as the existing docs for each method are clear that `False` is returned if the path is inaccessible for any reason.

**Comment by zooba** [comment]
> If you think it's important, I can include some conceptual remarks as you suggest.

I think it's helpful for people who come to our docs not just for reference, but to find out what they _should_ do. But I also can't come up with great wording that doesn't have so many qualifiers that you'd need to be an expert to understand it anyway.

Maybe it's a matter of phrasing it more like: "Returns `True` if the path refers to a file; otherwise, `False`. Paths that are inaccessible will return `False`, as will those that are not files. Use `exists` to determine whether the path is accessible, and `is_dir` or `is_link` to check for directories or symlinks."

Basically, "if you use this, here's what you'll get; if you actually wanted _this_, here's what you should do". But it's more of a rewrite, and maybe not worth your time or interest. I'm not going to force it, the current PR is fine.

**Comment by barneygale** [comment]
Makes sense! I'll experiment and see if I can make that sort of thing work. Ta!

File: Doc/library/pathlib.rst [hunk 1]
```diff
@@ -1068,11 +1077,9 @@ call fails (for example because the path doesn't exist).
...
 .. method:: Path.is_dir(*, follow_symlinks=True)
 
-   Return ``True`` if the path points to a directory, ``False`` if it points
-   to another kind of file.
-
-   ``False`` is also returned if the path doesn't exist or is a broken symlink;
-   other errors (such as permission errors) are propagated.
+   Return ``True`` if the path points to an accessible directory, ``False``
+   otherwise. Use :meth:`Path.exists` to test if a path is accessible without
+   checking its type.
```
Comment (zooba) [inline Doc/library/pathlib.rst]:
```suggestion
   Return ``True`` if the path points to an accessible directory.
   Use :meth:`Path.exists` to determine if a path is not accessible,
   or the other ``is_*`` methods to test for other types.
```

How about this? Looking at it again, "False otherwise" seems redundant, and I think suggesting other `is_*` checks really emphasises that "`not is_dir()` doesn't imply `is_file()`".

File: Doc/library/pathlib.rst [hunk 2]
```diff
@@ -1123,46 +1128,37 @@ call fails (for example because the path doesn't exist).
 
 .. method:: Path.is_symlink()
 
-   Return ``True`` if the path points to a symbolic link, ``False`` otherwise.
-
-   ``False`` is also returned if the path doesn't exist; other errors (such
-   as permission errors) are propagated.
+   Return ``True`` if the path points to an accessible symbolic link,
```
Comment (zooba) [inline Doc/library/pathlib.rst]: Might be worth emphasising that this function does not follow symlinks? It's the odd one out, I believe.

File: Doc/library/pathlib.rst [hunk 3]
```diff
@@ -1123,46 +1128,37 @@ call fails (for example because the path doesn't exist).
...
 
 
 .. method:: Path.is_socket()
 
-   Return ``True`` if the path points to a Unix socket (or a symbolic link
-   pointing to a Unix socket), ``False`` if it points to another kind of file.
-
-   ``False`` is also returned if the path doesn't exist or is a broken symlink;
-   other errors (such as permission errors) are propagated.
+   Return ``True`` if the path points to an accessible Unix socket, ``False``
```
Comment (zooba) [inline Doc/library/pathlib.rst]: Okay, I'm starting to regret the "accessible" at this point 😆 But I think I've figured out why:

"Accessible" is really a property of the _path_, but these descriptions make it sound like it's a property of the _file_ (or in this case, the socket). And "inaccessible" can be ambiguous for some of these (imagine a removable drive, where the _path_ may be accessible enough to determine the type, but the _drive_ is not accessible if nothing is in there).

So new proposal (for all of them):
> Return True if the path points to a Unix socket. False will be returned if the path is invalid, inaccessible or missing, or if it points to something other than a Unix socket. Use Path.exists to distinguish between these cases.

File: Doc/library/pathlib.rst (hunk 1 above)
Comment (zooba) [reply Doc/library/pathlib.rst]: Added a different suggestion further down (or up... depending on which way the review reads...)

File: Doc/library/pathlib.rst (hunk 3 above)
Comment (nineteendo) [reply Doc/library/pathlib.rst]: Isn't that a bit long? How about:

> Return True if the path is accessible and points to a valid Unix socket...

File: Doc/library/pathlib.rst (hunk 2 above)
Comment (barneygale) [reply Doc/library/pathlib.rst]: I added a bit about it returning `True` for broken links, which I think is a slightly more practical spin on the same idea. What do you think?

File: Doc/library/pathlib.rst (hunk 3 above)
Comment (zooba) [reply Doc/library/pathlib.rst]: I'm really trying to get the reference to using Path.exists in there, to really clearly lead people towards that if they care about the different results (and thereby imply that "is_*" are for when you _don't_ care about the different results).

So while I do like the shorter phrasing, I think you then end up with all the same words describing the alternative ("Use Path.exists to distinguish between cases where the path is invalid, inaccessible, missing, or points to something other ..."). Given the choice, I'd rather more clearly explain _this_ function than the alternative (or not clearly explaining either of them).

Finding a balance between reference, specification, and teaching is the toughest part of documentation, and these docs have to do all three.

**Review by zooba — APPROVED** [review]
Love it. But do consider other feedback as well, if there's something that can make it even more approachable.

File: Doc/library/pathlib.rst (hunk 3 above)
Comment (nineteendo) [reply Doc/library/pathlib.rst]: And this? If the path is accesible, it's also not missing.

> Return True if the path is accessible and points to a valid Unix socket. Use Path.exists to not check for the latter.

Comment (zooba) [reply Doc/library/pathlib.rst]: It's not just not checking the type, it's also going to raise different errors based on why it's inaccessible. But I didn't want to go as far as suggesting to handle `PermissionError` vs `FileNotFoundError` vs whatever other OSErrors may be raised. 

The main thing to imply is that a `True` result from these functions implies one case, while a `False` result implies one of many cases, to prevent people thinking that `not is_file()` is the same as `is_dir()` (which is obvious to us, who are neck deep in this stuff, but is not otherwise obvious at first sight). Suggesting that `exists` will let you distinguish exactly what `False` means is the best way I've come up with so far to communicate it without making it too hard to understand.

Also, I think "Use Path.exists to not check for the latter" is very difficult to parse. But even reworded for clarity, I don't think it would communicate what I'm wanting to communicate here.

Comment (nineteendo) [reply Doc/library/pathlib.rst]: You mean for Windows? Because for Unix the only difference is checking the file type:

https://github.com/python/cpython/blob/a7f83721d97f7b2b6c0efb21929b359c4143ed22/Lib/pathlib/_abc.py#L431-L435

https://github.com/python/cpython/blob/a7f83721d97f7b2b6c0efb21929b359c4143ed22/Lib/pathlib/_abc.py#L525-L528

Comment (zooba) [reply Doc/library/pathlib.rst]: Oh, does `exists` suppress all exceptions as well? If that's the case, we should be suggesting calling `stat` instead @barneygale

Comment (nineteendo) [reply Doc/library/pathlib.rst]: > Return True if the path points to a Unix socket. False is returned if the path is invalid, inaccessible, missing or points to something else. Use Path.stat to distinguish between these cases.

Comment (barneygale) [reply Doc/library/pathlib.rst]: Thanks for the feedback, both.

@nineteendo I did try the "something else" formulation, but being a bit more explicit is helpful for readers.

**Comment by barneygale** [comment]
Thanks very much for the review!

File: Doc/library/pathlib.rst [hunk 4]
```diff
@@ -1068,11 +1079,10 @@ call fails (for example because the path doesn't exist).
...
 
-   Return ``True`` if the path points to a directory, ``False`` if it points
-   to another kind of file.
-
-   ``False`` is also returned if the path doesn't exist or is a broken symlink;
-   other errors (such as permission errors) are propagated.
+   Return ``True`` if the path points to a directory. ``False`` will be
+   returned if the path is invalid, inaccessible or missing, or if it points
+   to something other than a directory. Use :meth:`Path.stat` to distinguish
+   between these cases.
```
Comment (nineteendo) [inline Doc/library/pathlib.rst:1085]: Is it necessary to clarify this?

```suggestion
   returned if the path is invalid, inaccessible or missing, or if it points
   to something else. Use :meth:`Path.stat` to distinguish between
   these cases.
```

You could still make it shorter in that case:

```suggestion
   returned if the path is invalid, inaccessible or missing, or if it doesn't
   point to a directory. Use :meth:`Path.stat` to distinguish between these
   cases.
```

Comment (barneygale) [reply Doc/library/pathlib.rst:1085]: https://github.com/python/cpython/pull/118243#discussion_r1600416339
